FE-feat-atoms-jjy 이슈#16 atomic 요소 개발 완료

JunYoung7 opened the pull request:
### 📕 Issue Number

Close #16 
<br/>

### 📗 작업 내역

> 구현 내용 및 작업 했던 내역

- hr/Line atom 구현 
![image](https://user-images.githubusercontent.com/61405355/100358149-1778e480-3039-11eb-94cb-dfbbe220185b.png)


- div atoms 작업

   - ToggleButton 구현
![image](https://user-images.githubusercontent.com/61405355/100358351-6d4d8c80-3039-11eb-9616-7bd985670083.png)

   - TextToggleButton 구현
![image](https://user-images.githubusercontent.com/61405355/100358397-79d1e500-3039-11eb-8366-2cc92f901c88.png)

   - ColorLabel 구현
![image](https://user-images.githubusercontent.com/61405355/100358427-85251080-3039-11eb-831c-7f54ab402cb8.png)

<br></br>

- p atom 작업
   - UnderlineText 작업
![image](https://user-images.githubusercontent.com/61405355/100358485-9a9a3a80-3039-11eb-8796-08cae7911885.png)

<br/>

### 📘 작업 유형

- 신규 기능 추가

<br/>

### 📝 PR 특이 사항

> 코드를 작성하면서 알아가는 것들도 있어서 놓치거나 부족한 점이 많을 수 있습니다..!! 혹시 발견하신다면 피드백 주시면 정말 감사하겠습니다😺

- toggle button의 색상 변화 부분에 대해, 제가 접근한 방식과 다르게 구현하는 방법들에 대해서도 궁금합니다!

<br/><br/>

JunYoung7 committed:
feat : UnderlineText 구현

# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명
- 밑줄 친 text 재사용을 위한 underlineText 구현
- text에 밑줄을 넣기 위해 text-decoration : underline 적용

JunYoung7 committed:
feat : ToggleButton 구현

- toggle 할 수 있는 버튼을 사용하기 위한 ToggleButton 구현
- 버튼 선택 여부를 알고 있기 위한 isLeftSelected state 변수 사용
- 색깔 변화 함수와 props onClick함수 동시 호출을 위한 onClick에 복수 함수 연결
# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명

JunYoung7 committed:
feat : TextToggleButton 구현

- text로 이루어진 toggle button을 만들기 위한 textToggleButton 구현
- 투명하고 테두리 없는 텍스트 버튼을 위해 만들어 놓은 textButton 재사용
# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명

JunYoung7 committed:
refactor : ToggleButton, RoundShortButton 수정

- roundShortButton에 border를 주기 위한 border solid 속성, 크기 지정
- ToggleButton 크기 수정을 위한 defaultProps 값 수정
- toggleButton text 변화 확인을 위한 toggleButton.stories.tsx 파일 수정
# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명

JunYoung7 committed:
feat : ColorLabel 구현

- 선택된 색상 표시를 위한 ColorLabel 구현
- 재사용을 위한 색상, width, height 입력할 수 있도록 구현
# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명

leejiwoo2002 commented on client/src/components/atoms/div/ColorLabel/ColorLabel.tsx:
```diff
@@ -0,0 +1,33 @@
+const Div = styled.div<Props>`
+  background-color: ${(props) => props.backgroundColor};
+  width: ${(props) => props.width};
+  height: ${(props) => props.width};
```
width와 height를 공유하는, 즉 정사각형 모양의 컴포넌트일때 이런 상황이 만들어지는데요.
저는 size라고 명명해서 하나의 props로 받아오는데, 더 좋은 방법이 있을지 의논해보고 싶습니다

JunYoung7 replied:
하나로 받는 것도 좋을 것 같아요! 피드백 감사드립니다

do02reen24 commented on client/src/components/atoms/div/ColorLabel/ColorLabel.stories.tsx:
```diff
@@ -0,0 +1,32 @@
+  component: ColorLabel,
+  argTypes: {
+    width: { control: 'width' },
+    height: { control: 'height' },
```
contorl로 정의할 수 있는 속성으로 [이것들](https://storybook.js.org/docs/react/essentials/controls#annotation)만 있다고 알고 있었는데 잘 동작하는지 한 번 확인해주시면 좋을 것 같아요 ㅎㅎ

JunYoung7 replied:
앗, 피드백해주셔서 감사합니다. 내일 바로 반영하도록 하겠습니다. 😃

do02reen24 commented on client/src/components/atoms/div/ColorLabel/ColorLabel.stories.tsx:
```diff
@@ -0,0 +1,32 @@
+export default {
+  title: 'Atoms/div/ColorLabel',
+  component: ColorLabel,
+  argTypes: {
```
`backgroundColor: { control: 'color' }` 로 정의해주어 storybook에서 색깔을 자유롭게 바꿔볼 수 있도록 해주시면 더 좋을 것 같아요!😀

JunYoung7 replied:
넵 내일 한번 적용해보도록 하겠습니다. 감사합니다!! 😃

do02reen24 reviewed (commented):
ToggleButton이 구현하기 까다로웠을텐데 수고하셨습니다😀👍🔥

Zigje9 commented on client/src/components/atoms/div/ToggleButton/ToggleButton.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+  onClickRight: () => void;
+}
+
+export default {
+  title: 'Atoms/div/ToggleButton',
+  component: ToggleButton,
+  argTypes: {
+    leftButtonName: { control: 'leftButtonName' },
+    rightButtonName: { control: 'rightButtonName' },
+  },
+};
+
+export const toggleButton = ({ ...args }: Props): JSX.Element => {
+  return <ToggleButton {...args} />;
+};
+
+toggleButton.story = {
+  name: 'ToggleButton',
+};
+
+toggleButton.args = {
+  leftButtonName: '수입',
+  rightButtonName: '지출',
+  onClickLeft: undefined,
+  onClickRight: undefined,
+  width: '200px',
+  height: '50px',
+};
```
스토리북의 컨트롤 기능을 잘 해두신 것 같습니다. 저도 다음부터는 적용해 봐야 겠네요!

JunYoung7 replied:
도연님이 알려주셨는데 control은 가능한 요소가 [이것들](https://storybook.js.org/docs/react/essentials/controls#annotation)만 있다고 하네요! 함께 참고하셔도 좋을 것 같습니다

Zigje9 commented on client/src/components/atoms/div/TextToggleButton/TextToggleButton.tsx:
```diff
@@ -0,0 +1,86 @@
+}: Props) => {
+  const [isleftSelected, setIsleftSelected] = useState(true);
+
+  useEffect(() => {
+    if (isleftSelected === false) {
+      leftColor = myColor.primary.darkGray;
+      rightColor = myColor.primary.accent;
+    } else {
+      leftColor = myColor.primary.accent;
+      rightColor = myColor.primary.darkGray;
+    }
+  }, [isleftSelected]);
+
+  const changeColor = () => {
+    setIsleftSelected(!isleftSelected);
+  };
+
+  return (
+    <TextToggleButtonContainer {...args}>
+      <Button
+        color={leftColor}
+        onClick={() => {
+          changeColor();
+          onClickLeft();
+        }}
+      >
+        {leftButtonName}
+      </Button>
+      <Button
+        color={rightColor}
+        onClick={() => {
+          changeColor();
+          onClickRight();
+        }}
+      >
+        {rightButtonName}
+      </Button>
+    </TextToggleButtonContainer>
+  );
+};
```
onClickRight() 와 onClickLeft() 는 어떤 역할을 하는건가요??

JunYoung7 replied:
예를 들어 수입 버튼과 지출 버튼이 있다고 가정하겠습니다. 이 경우에, 수입 버튼이 눌렸을 때 발생할 이벤트와 지출 버튼이 눌렀을 때 발생할 이벤트 각각을 분리해서 props로 받도록 한 것 입니다. 하나의 함수로 처리하는 방법도 있을 것 같아요!

Zigje9 replied:
아하 수입과 지출 선택했을때 반영되어야 하는 것을 고려하신 거군요 👍

Zigje9 reviewed (approved):
고생하셨습니다 👍

JunYoung7 committed:
refactor : ColorLabel, ToggleButton stories 수정

- control 요소를 제대로 활용하기 위한 type 재지정
# <Why>를 위한 <What>
# 로그인을 위한 OAuth api 구현
# 객체 의존성 역전을 위한 프로토콜 추상화

# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# --------------------------------------------------------------
# 메세지는 한글로 작성
# 타입 첫 문자는 소문자로 작성
# 제목의 끝은 명사형으로 작성
# 제목 끝에 마침표(.) 금지
# 본문은 "어떻게" 보다 "무엇을", "왜"를 설명
